keps/sig-cluster-lifecycle/kubeadm/20190506-kubeadm-machine-output.md

```diff
@@ -0,0 +1,363 @@
+
+## Proposal
+
+### User Stories
+
+The examples in the user stories are predicated on the existence of:
+
+* [A friendly bootstrap token struct](#a-friendly-bootstrap-token-struct)
+* [Additional functions for Go templates](#go-template-functions)
+
+#### Story 1
+
+The deployment of a multi-node Kubernetes cluster is automated by parsing [the JSON output](#kubeadm-init-json-output) of `kubeadm init` for the emitted `kubeadm join` command.
+
+1. Run `kubeadm init -o json | jq -r '"kubeadm join \(.node0) --token \(.token.id) --discovery-token-ca-cert-hash \(.caCrt)"'`
+
+#### Story 2
+
+A script requires a list of token IDs.
+
+1. Run `kubeadm token list -o '{{range .}}{{println .ID}}{{end}}'`
+
+#### Story 3
+
+A script returns a list of token IDs as they are discovered.
+
+1. Run `kubeadm token list --stream -o '{{println .ID}}'`
+
+#### Story 4
+
+A script needs to process the IDs of all the non-expired tokens:
+
+1. Run `kubeadm token list -o json | jq '.[] | select(.expires | fromdate > now) | .id'`
+
+#### Story 5
+
+A script needs to process the IDs of all the non-expired tokens as the tokens are discovered:
+
+1. Run `kubeadm token list --stream -o json | jq 'select(.expires | fromdate > now) | .id'`
+
+### Implementation Details/Notes/Constraints
+
+#### Details
+
+##### A centralized printer system
+The first design detail is related to how commands handle output...they don't. Or rather, they shouldn't. A command should perform an action and return a result, not print that result. Much like the package [`k8s.io/cli-runtime/pkg/printers`](https://github.com/kubernetes/cli-runtime), a package dedicated to printing will be introduced to kubeadm.
+
+Whether by virtue of an interface that describes a printer, or a single, exported function, the signature of a printer should be:
+
+```golang
+func Print(w io.Writer, format string, data interface{}) error
+```
+
+A very pseudo-codish implementation of the above signature might look something like this:
+
+```golang
+func Print(w io.Writer, format string, data interface{}) error {
+	// If data is an io.Reader then copy the reader to the writer.
+	if r, ok := data.(io.Reader); ok {
+		_, err := io.Copy(w, r)
+		return err
+	}
+
+	// Handle pre-defined formats.
+	switch format {
+	case "json":
+		return json.NewEncoder(w).Encode(data)
+	case "yaml":
+		buf, err := yaml.Marshal(data)
+		if err != nil {
+			return err
+		}
+		_, err := io.Copy(w, bytes.NewReader(buf))
+		return err
+	case "text":
+		_, err := fmt.Fprintln(w, data)
+		return err
+	}
+
+	// Treat the format as a Go template.
+	tpl, err := template.New("t").Parse(format)
+	if err != nil {
+		return err
+	}
+	return tpl.Execute(w, data)
+}
+```
+
+##### Decoupling commands from printing
+If a command is no longer in charge of rendering its output, how does a command submit its output to be rendered? There are several possibilities. Commands could:
```

Comment:
Afaik, current kubeadm output generation is spread across the entire codebase and deep tree of nested calls (sigh)
Therefore, I consider this is part the most tricky of the entire proposal.

One point that should be considered is the potential impact of the different solutions on the long term ambition to make kubeadm more vendor friendly

Reply:
Hi @fabianofranz,

> One point that should be considered is the potential impact of the different solutions on the long term ambition to make kubeadm more vendor friendly

Good point. I don't think it should be too difficult to account for vending. Whether commands receive a print queue or submit jobs to an exported print queue, the back-end of the queue can be something that defaults to one listener (that doesn't use cli-runtime) and can be set to a cli-runtime printer by the kubeadm CLI code. This way if someone vendors just the app, they won't pull in cli-runtime.

Reply:
ok, let's keep this in mind; fyi latest discussions on kubeadm/vendoring are focused only on the API/component config, so basically ATM this can be reduced to a simple concern about code design